packages/hardhat/contracts/MyUSD/CoinEngine.sol

```diff
@@ -0,0 +1,234 @@
+            totalDebt += userInterest;
+        }
+        s_userLastUpdateTime[user] = block.timestamp;
+    }
+
+    // Allows users to add collateral to their account
+    function addCollateral() public payable {
+        if (msg.value == 0) {
+            revert Engine__InvalidAmount(); // Revert if no collateral is sent
+        }
+        s_userCollateral[msg.sender] += msg.value; // Update user's collateral balance
+        emit CollateralAdded(msg.sender, msg.value, i_ethPriceOracle.price()); // Emit event for collateral addition
+    }
+
+    // Allows users to withdraw collateral as long as it doesn't make them liquidatable
+    function withdrawCollateral(uint256 amount) external {
+        if (amount == 0 || s_userCollateral[msg.sender] < amount) {
+            revert Engine__InvalidAmount(); // Revert if the amount is invalid
+        }
+
+        // Temporarily reduce the user's collateral to check if they remain safe
+        uint256 newCollateral = s_userCollateral[msg.sender] - amount;
+        s_userCollateral[msg.sender] = newCollateral;
+
+        // Validate the user's position after withdrawal
+        if (s_userMinted[msg.sender] > 0) {
+            _validatePosition(msg.sender);
+        }
+
+        // Transfer the collateral to the user
+        payable(msg.sender).transfer(amount);
+
+        emit CollateralWithdrawn(msg.sender, msg.sender, amount, i_ethPriceOracle.price()); // Emit event for collateral withdrawal
+    }
+
+    // Allows users to mint stablecoins based on their collateral
+    function mintStableCoin(uint256 mintAmount) public {
+        if (mintAmount == 0) {
+            revert Engine__InvalidAmount(); // Revert if mint amount is zero
+        }
+        _accrueInterest();
+        _updateUserDebt(msg.sender);
+        s_userMinted[msg.sender] += mintAmount;
+        totalDebt += mintAmount;
+        _validatePosition(msg.sender);
+        bool success = i_myUSD.mintTo(msg.sender, mintAmount);
+        if (!success) {
+            revert Engine__MintingFailed(); // Revert if minting fails
+        }
+    }
+
+    // Allows users to burn stablecoins and reduce their debt
+    function burnStableCoin(uint256 burnAmount) public {
+        if (burnAmount == 0 || burnAmount > s_userMinted[msg.sender]) {
+            revert Engine__InvalidAmount(); // Revert if burn amount is invalid
+        }
+        _accrueInterest();
+        _updateUserDebt(msg.sender);
+        s_userMinted[msg.sender] -= burnAmount;
+        totalDebt -= burnAmount;
+        bool success = i_myUSD.burnFrom(msg.sender, burnAmount);
+        if (!success) {
+            revert Engine__BurningFailed(); // Revert if burning fails
+        }
+    }
+
+    // Retrieves the user's position, including minted amount and collateral value
+    function _getUserPosition(address user) private view returns (uint256 mintedAmount, uint256 collateralValue) {
+        mintedAmount = s_userMinted[user]; // Get user's minted amount
```

Comment:
Seems like we need to add the amount of new debt from the interest rate when calculating the user's position. Not actually adding it since this is a view method but don't we need to calculate how much additional debt has accrued? For instance the interest rate may have changed a long time ago but the borrower has never minted or burned more tokens so the new interest rate doesn't actually impact them until they do one of those actions. I think the only issue is that we can't actually tell if they are in a liquidatable state.